resources/session03/calculator.py

```diff
@@ -0,0 +1,169 @@
+  </div>
+  <button type="submit" formaction="add" class="btn btn-default">Add</button>
+  <button type="submit" formaction="subtract" class="btn btn-default">Subtract</button>
+  <button type="submit" formaction="multiply" class="btn btn-default">Multiply</button>
+  <button type="submit" formaction="divide" class="btn btn-default">Divide</button>
+</form>
+"""
+
+my_template = """\
+<h3>{a} {verb} {b} equals {c}</h3>
+<a class="btn btn-primary" href="/" role="button">Again?</a>
+"""
+
+error_template = """\
+<h3>OMG SOMETHING HAPPENED</h3>
+<p>{error}</p>
+<a class="btn btn-primary" href="/" role="button">Go Home</a>
+"""
+
+
+# Here lies the application section.  Normally would be a module someplace else
+def render(template, **kwargs):
+    '''In a real app, this would be built to allow inheritance
+       between templates.  It's late, and I'm cheating.'''
+    template = layout.format(input=template)
+    return template.format(**kwargs)
+
+
+class Erlenmeyer:
+    '''Sorry for the pun'''
+    # dict for mapping urls to views
+    paths = {}
+
+    def add_path(self, path, f):
+        self.paths[path] = f
+
+    def route(self, path):
+        '''Decorator to add view functions.  Way more complex than needed
+           for this trivial case, but this made it interesting :-p'''
+        def decorated(f):
+            self.add_path(path, f)
+            return f
+        return decorated
+
+    def _parse_qs(self, query):
+        '''Return dict built from query string parameters'''
+        d = {}
+        if '&' in query:  # multiple params
+            for item in query.split('&'):
+                a, b = item.split('=')
+                d.update({a: b})
+        elif '=' in query:  # one param
+                a, b = item.split('=')
+                d.update({a: b})
+        return d   # will be empty with no params
+
+    def reply(self, start_response, response_body, status='200 OK'):
+        '''Notifies wsgi server of output and returns list of output'''
+        response_headers = [('Content-Type', 'text/html'),
+                            ('Content-Length', str(len(response_body)))]
+        start_response(status, response_headers)
+        return [response_body.encode('utf8')]
+
+    def __call__(self, environ, start_response):
+        '''Returns content to user'''
+
+        path = environ['PATH_INFO']
+        qdict = self._parse_qs(environ['QUERY_STRING'])
+
+        try:
+            response_body = self.paths[path](**qdict)
+        except KeyError:
+            status = '404 Not Found'
+            response_body = render(error_template, error=status)
+            return self.reply(start_response, response_body, status)
+        except TypeError or ValueError:
+            status = '400 Bad Input'
+            response_body = render(error_template, error=status)
+            return self.reply(start_response, response_body, status)
+
+        return self.reply(start_response, response_body)
```

Comment:
You've called `self.reply` three times here.  Do you really need to?  Could you do it only the once?